### ISSUE TITLE
[1.3.1.2/1.3.2/1.4] Village gen weirdness

### ISSUE THREAD

**Issue by thephoenixlodge** [Ivorforce/RecurrentComplex#221]
Something in 1.3.1.2 seems to have broken most of my structure generation in villages. The bizarre thing is that two of my structures, and some of the default structures gen, but every village spawns as a large network of paths with only a few buildings.

In this pic, the village towards the top was generated on 1.3.1.2, while the one below was genned on 1.3.1.1. I first noticed the issue while running 1.3.2.
http://imgur.com/a/SeaqI

**Comment by Ivorforce** [comment]
Try using just 1.3.2 and seeing if anything is broken there. 1.3.1.2 and earlier had known issues but the latest version should be completely stable in that regard.

**Comment by thephoenixlodge** [comment]
As I said, I first noticed the issue while I was running 1.3.2

**Comment by thephoenixlodge** [comment]
Can also confirm that the issue still exists in 1.4.0
(freshly generated village)
![2017-06-18_19 48 57](https://user-images.githubusercontent.com/5405605/27259170-a30c0918-5460-11e7-8b9d-0cf1e3e6fbc0.png)

**Comment by ShicKla** [comment]
I can also confirm this issue in 1.3.2 and 1.4, except in my case non of my custom village buildings are generating. Its almost as if the generated is either failing to look for active structures or just ignoring them.

**Comment by ShicKla** [comment]
Update:
I think I found the reason as to why my structures weren't spawning. It seems that if they don't have the Rotatable flag as true they seem to be ignored. As soon as I made my structures rotatable they started spawning. Sadly I intentionally made them non-rotatable because they have modded blocks that wont rotate.

@thephoenixlodge can you confirm with your custom structures if they do the same?

**Comment by Ivorforce** [comment]
Yes, that's quite possible. If structures are not rotatable, the village tries to expand forever to find a place to spawn for those buildings, but they will only take a place where the road is, say, to the south. They shouldn't be _entirely_ ignored though - if they are that's a bug. They should just not be able to spawn quite as often and might cause road networks.

**Comment by thephoenixlodge** [comment]
yea I can't rotate mine due to the way I spawn NPCs, and also usage of C&B.

**Comment by thephoenixlodge** [comment]
Comparatively, I can't stick with 1.3.1.1 either due to it not spawning any structures set to face anything other than north.

**Comment by ShicKla** [comment]
I did figure out a clever work around though. I create and simple structure with and adequate size footprint and put a spawner in the middle that will then spawn the desired structure without rotating it. Sure this means there is a chance the structure will face away from the road, but I will take that over not having the structure at all.

**Comment by thephoenixlodge** [comment]
Would it not be a bug that I'm only getting ~5 buildings per village with this though?

**Comment by ShicKla** [comment]
@thephoenixlodge
Is it ~5 buildings total or ~5 non-vanilla?

**Comment by Hootling** [comment]
To add to this, I am getting this even after disabling Recurrent Complex.   Huge networks of pathways with tons of "light posts" like in the picture above.  Even though mod is disabled, it's changed something permanently somewhere, and I don't know where to fix it.

**Comment by ShicKla** [comment]
@Hootling
Does the rampant pathways happen if you make a brand new world without RC installed, or is it only happening on a world that was created with RC installed?

**Comment by Hootling** [comment]
It happened even on new worlds without RC installed.   I checked this several times to make sure I wasn't missing anything.  I even tried Curse's "repair minecraft" feature, no luck.

I had to make a new Curse profile, and copy over my modlist and configs to fix.  So whatever it changes, it's not in my /mods  /config folders.

**Comment by Ivorforce** [comment]
@thephoenixlodge C&B should be compatible last time I tested. I'm not sure how you spawn in NPCs but if possible I'm willing to add new features to be able to support the rotation of your structure :)
Also, what do you mean by '5 structures per village'? With what settings and which structure?

@ShicKla Clever. I'm considering adding the possibility for a compatibility mode though where the structure opens up every side as an 'entrance' and thus spawns easily without being rotatable.

@Hootling that seems impossible to me. ReC doesn't change anything permanently anywhere. You cannot 'disable' the mod though - you need to close MC, delete it from the mods folder and restart it.

**Comment by Hootling** [comment]
It seemed impossible to me too.   At first I disabled it (trying to see if it was indeed the mod) that did not work so I deleted it entirely.  Still same problem.   

I had to rebuild my Curse profile to fix it.   If I have time I'll make a copy of my profile and see if I can replicate.  

Only thing I can think of is that maybe there's some interaction with one of my other mods.   I have no idea.  But I definitely deleted it.

**Comment by Ivorforce** [comment]
This commit should solve all problems related to this issue. Non-rotatable fixed front buildings will continue to expand villages very far since that's how vanilla behaves (unless I made another mistake somewhere which strengthens the problem).
Just select 'All' as front for non-rotatable structures from the next version on.

@Hootling  if you run into this problem again or find out anything else please create a new issue since that would be something different. :)

**Comment by thephoenixlodge** [comment]
Sorry for the lack of response for a while here. Due to reasons, I spawn the NPCs in my structures by having a script block spawn a secondary structure with a redstone circuit which spawns the NPC with a command block and then clears itself. Only way I can think that would allow me to make structures that want NPCs to be in specific spots rotatable would be for structures spawned with script blocks able to inherit the rotation of the structure the script block is in.

As for the C&B compat, yes it is compatible and works well, saving the data of the blocks correctly, but they don't get rotated with the structure, so any sort of C&B design that is directional doesn't look right at anything bar the original orientation. Same thing happens with Architecture Craft blocks.

And finally, as for the ~5 structures, take a look at the screenshot above, It varied from village to village, but I was hardly even getting any of the vanilla buildings - mostly just ones added by other mods.

**Comment by Ivorforce** [comment]
> with a redstone circuit which spawns the NPC with a command block and then clears itself

That sounds unnecessarily complicated. Why not just use a command spawn script?

> As for the C&B compat, yes it is compatible and works well, saving the data of the blocks correctly, but they don't get rotated with the structure, so any sort of C&B design that is directional doesn't look right at anything bar the original orientation. Same thing happens with Architecture Craft blocks.

I'd report that to the mod authors. The ability to rotate a Tile Entity is, at this point, a vanilla capability and not something just ReC expects.

> And finally, as for the ~5 structures, take a look at the screenshot above, It varied from village to village, but I was hardly even getting any of the vanilla buildings - mostly just ones added by other mods.

ReC doesn't change the number of other structures that spawn in villages per se. It's possible having more structures registered has an effect on the quantity though. That's to be expected - you can change the balancing with this in the config though.

**Comment by thephoenixlodge** [comment]
> That sounds unnecessarily complicated. Why not just use a command spawn script?
Custom NPCs is very much designed to work in prebuilt maps - all it's data is stored within the world folder. To help me work around this, joshiejack made a mod called copypaste which copies whatever is placed into /config/copy into the world save folder when the world is loaded, which is how I get my NPC clone data into every world, as required for the NPC clone spawn command to work.
Having a script block spawn a clone in a structure directly does work, kinda, but as it turns out, somehow this happens further down the chain of events than initial generation of a newly created world, so directly having a script block in the base structure spawn the NPC didn't work for chunks genned as the world is initially created - the clone data doesn't exist yet for the command to spawn the NPC from. So the structures in those chunks would gen just fine, and the script blocks would activate, but there'd be no NPCs in them in those chunks.
By having the sub structure with the redstone circuit, activation of the command block to run the npc spawn command is forcibly delayed until the player is actually in the world, guaranteeing the data has copied across for the NPC to be spawned correctly.

TL;DR: Wacky workaround to a bizarre issue caused by me forcing the mods to work to my intended vision.

**Comment by Ivorforce** [comment]
Aha, that does sound complicated.
Maybe you could ask the author of custom NPCs to add an option that allows NPCs to store their data within the NPC instead? (maybe just once?)

**Comment by Razzapan-Deathpunch** [comment]
HEY, i love your modpack, but im haviung issues finding actual servers for it. i've watched the others playing on a sprout server but i can't actually find any way to find an active list of them? nor can i find the server files. am i looking in teh wrong places
